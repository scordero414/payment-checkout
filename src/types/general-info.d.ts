import { AlertColor } from '@mui/material';

export interface GeneralInfoState {
  isLoading: boolean;
  alertMessage: string;
  isAlertActive: boolean;
  alertType: AlertColor;
}
