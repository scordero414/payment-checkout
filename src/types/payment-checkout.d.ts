export interface PaymentCheckoutState {
  value: string;
}
