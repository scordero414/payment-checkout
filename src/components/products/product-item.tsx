import { Product } from '@/types/products';
import {
  Card,
  CardActionArea,
  CardContent,
  CardMedia,
  Grid,
  Typography,
} from '@mui/material';
import { useRouter } from 'next/navigation';

interface ProductItemProps {
  product: Product;
}

export const ProductItem = ({ product }: ProductItemProps) => {
  const { title, image, id, price } = product;

  const router = useRouter();

  const handleOpenProduct = () => {
    router.push(`/product/${id}`);
  };
  return (
    <Grid item container xs={12} sm={5} md={3} lg={2} xl={1.5}>
      <Card sx={{ width: '100%', maxHeight: 320 }} onClick={handleOpenProduct}>
        <CardActionArea>
          <CardMedia
            component="img"
            height={200}
            image={image}
            sx={{ objectFit: 'contain' }}
            alt={`product-image-${id}`}
          />
          <CardContent>
            <Typography gutterBottom variant="h6" component="div" noWrap>
              {title}
            </Typography>
            <Typography variant="body2" color="text.secondary">
              {`$${price}`}
            </Typography>
          </CardContent>
        </CardActionArea>
      </Card>
    </Grid>
  );
};
