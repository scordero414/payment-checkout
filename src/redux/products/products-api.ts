import { sliceNamesConstants } from '@/constants/slice-names-constants';
import { ProcessPaymentData } from '@/types/payment-checkout';
import { Product } from '@/types/products';
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

export const productsApi = createApi({
  reducerPath: sliceNamesConstants.productsApi,
  baseQuery: fetchBaseQuery({
    baseUrl: process.env.NEXT_PUBLIC_PRODUCTS_BASE_URL,
  }),
  endpoints(builder) {
    return {
      getProducts: builder.query<Product[], void>({
        query: () => '/products',
      }),
      processPayment: builder.mutation<unknown, ProcessPaymentData>({
        query: params => ({
          url: `/products/${params.product.id}`,
          method: 'PUT',
          body: params.product,
        }),
      }),
      processBadPayment: builder.mutation<unknown, ProcessPaymentData>({
        query: params => ({
          url: `/products/${params.product.id}`,
          method: 'POST',
          body: params,
        }),
      }),
    };
  },
});

export const {
  useGetProductsQuery,
  useProcessBadPaymentMutation,
  useProcessPaymentMutation,
} = productsApi;
