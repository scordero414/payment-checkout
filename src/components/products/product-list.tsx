'use client';

import { ProductItem } from '@/components/products/product-item';
import { useGetProductsQuery } from '@/redux/products/products-api';
import { Grid } from '@mui/material';

export const ProductList = () => {
  const { data: products = [] } = useGetProductsQuery();

  return (
    <Grid
      item
      container
      pt={10}
      gap={{ xs: 2, md: 4 }}
      sx={{ justifyContent: 'center' }}>
      {products.map(product => (
        <ProductItem product={product} key={product.id} />
      ))}
    </Grid>
  );
};
