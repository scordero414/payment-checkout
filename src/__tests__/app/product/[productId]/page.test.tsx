import ProductPage from '@/app/product/[productId]/page';
import { render, waitFor } from '@testing-library/react'; // Update the path as needed
import { notFound } from 'next/navigation';
import fetchMock from 'jest-fetch-mock';
import '@testing-library/jest-dom';
import { Product } from '@/types/products';

jest.mock('next/navigation', () => ({
  notFound: jest.fn(),
}));

fetchMock.enableMocks();

describe('ProductPage', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it('should render product details', async () => {
    const productId = '20';
    const product: Product = {
      id: 20,
      title: 'Test Product',
      description: 'Test Description',
      price: 110,
      image: 'https://fakestoreapi.com/img/61U7T1koQqL._AC_SX679_.jpg',
    };
    fetchMock.mockResponseOnce(JSON.stringify(product));

    const jsx = await ProductPage({ params: { productId } });

    const { getByText, container } = render(jsx);

    await waitFor(() => {
      expect(getByText(product.title)).toBeInTheDocument();
      expect(getByText(product.description)).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  it('should call notFound if product is not found', async () => {
    const productId = '2';
    fetchMock.mockResponseOnce(JSON.stringify({}));

    const jsx = await ProductPage({ params: { productId } });

    render(jsx);

    await waitFor(() => {
      expect(notFound).toHaveBeenCalled();
    });
  });
});
