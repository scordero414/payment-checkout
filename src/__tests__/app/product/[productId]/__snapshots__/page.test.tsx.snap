// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage should render product details 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXl css-1yk8vz4-MuiContainer-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-m9nwi6-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5.5 css-e87de7-MuiGrid-root"
      >
        <img
          alt="product-image"
          data-nimg="1"
          decoding="async"
          height="300"
          loading="lazy"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Ffakestoreapi.com%2Fimg%2F61U7T1koQqL._AC_SX679_.jpg&w=3840&q=75 3840w"
          style="color: transparent; width: 100%; height: 100%; object-fit: contain; border-radius: 10px;"
          width="500"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5.5 css-110rxua-MuiGrid-root"
      >
        <div
          class="MuiStack-root css-nen11g-MuiStack-root"
        >
          <h1
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Test Product
          </h1>
          <h1
            class="MuiTypography-root MuiTypography-h6 css-tjvwag-MuiTypography-root"
          >
            $
            110
          </h1>
          <h1
            class="MuiTypography-root MuiTypography-subtitle1 css-10wpov9-MuiTypography-root"
          >
            Test Description
          </h1>
        </div>
        <div
          class="MuiBox-root css-uwwqev"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Pay with credit card
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="PrivateSwipeArea-root css-17j2vre"
            style="height: 20px;"
          />
          <div
            class="MuiBox-root css-188yn3u"
          >
            <div
              class="MuiBox-root css-pwmwbi"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-1nbvpvo-MuiTypography-root"
            >
              Checkout
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
